locator/tablets.cc

```diff
@@ -555,14 +556,26 @@ dht::token_range tablet_map::get_token_range_after_split(const token& t) const n
     return get_token_range(id_after_split, log2_tablets_after_split);
 }
 
-std::optional<tablet_replica> maybe_get_primary_replica(tablet_id id, const tablet_replica_set& replica_set, std::function<bool(const tablet_replica&)> filter) {
-    const auto replicas = replica_set | std::views::filter(std::move(filter)) | std::ranges::to<tablet_replica_set>();
+std::optional<tablet_replica> maybe_get_primary_replica(tablet_id id, const tablet_replica_set& replica_set, const locator::topology& topo, std::function<bool(const tablet_replica&)> filter) {
+    tablet_replica_set replica_set_copy = replica_set;
+    std::ranges::sort(replica_set_copy, [&topo] (const tablet_replica& a, const tablet_replica& b) {
+        const auto loc_a = topo.get_location(a.host);
+        const auto loc_b = topo.get_location(b.host);
+        if (loc_a.dc != loc_b.dc) {
+            return loc_a.dc < loc_b.dc;
+        }
+        if (loc_a.rack != loc_b.rack) {
+            return loc_a.rack < loc_b.rack;
+        }
+        return a.host < b.host;
```

Comment:
let's add endpoint_dc_rack::operator<=>(const endpoint_dc_rack&) and simply use projection here.

Reply:
I'd need also the host_id to make the comparison which `endpoint_dc_rack` doesn't seem to contain.

Reply:
Hmm, in practice, with rf-rack-valid we won't have multiple hosts in a rack, but that's right in principle.

In any case, I prefer ti have the comparison function defined as a method rather than open coding it here.

Reply:
added in the last version, thanks.